File: Source/Client/Forms/FrmEditor_Skill.vb
Imports System.IO
Imports System.Windows.Forms
Imports Core

Friend Class frmEditor_Skill

    Private Sub TxtName_TextChanged(sender As Object, e As EventArgs) Handles txtName.TextChanged
        Dim tmpindex As Integer
        tmpindex = lstIndex.SelectedIndex
        Skill(GameState.EditorIndex).Name = Trim$(txtName.Text)
        lstIndex.Items.RemoveAt(GameState.EditorIndex - 1)
        lstIndex.Items.Insert(GameState.EditorIndex - 1, GameState.EditorIndex & ": " & Skill(GameState.EditorIndex).Name)
        lstIndex.SelectedIndex = tmpindex
    End Sub

    Private Sub CmbType_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbType.SelectedIndexChanged
        Skill(GameState.EditorIndex).Type = cmbType.SelectedIndex
    End Sub

    Private Sub NudMp_ValueChanged(sender As Object, e As EventArgs) Handles nudMp.ValueChanged
        Skill(GameState.EditorIndex).MpCost = nudMp.Value
    End Sub

    Private Sub NudLevel_ValueChanged(sender As Object, e As EventArgs) Handles nudLevel.ValueChanged
        Skill(GameState.EditorIndex).LevelReq = nudLevel.Value
    End Sub

    Private Sub CmbAccessReq_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbAccessReq.SelectedIndexChanged
        Skill(GameState.EditorIndex).AccessReq = cmbAccessReq.SelectedIndex
    End Sub

    Private Sub CmbClass_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbJob.SelectedIndexChanged
        Skill(GameState.EditorIndex).JobReq = cmbJob.SelectedIndex
    End Sub

    Private Sub NudCast_Scroll(sender As Object, e As EventArgs) Handles nudCast.ValueChanged
        Skill(GameState.EditorIndex).CastTime = nudCast.Value
    End Sub

    Private Sub NudCool_Scroll(sender As Object, e As EventArgs) Handles nudCool.ValueChanged
        Skill(GameState.EditorIndex).CdTime = nudCool.Value
    End Sub

    Private Sub NudMap_Scroll(sender As Object, e As EventArgs) Handles nudMap.ValueChanged
        Skill(GameState.EditorIndex).Map = nudMap.Value
    End Sub

    Private Sub CmbDir_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbDir.SelectedIndexChanged
        Skill(GameState.EditorIndex).Dir = cmbDir.SelectedIndex
    End Sub

    Private Sub NudX_Scroll(sender As Object, e As EventArgs) Handles nudX.ValueChanged
        Skill(GameState.EditorIndex).X = nudX.Value
    End Sub

    Private Sub NudY_Scroll(sender As Object, e As EventArgs) Handles nudY.ValueChanged
        Skill(GameState.EditorIndex).Y = nudY.Value
    End Sub

    Private Sub NudVital_Scroll(sender As Object, e As EventArgs) Handles nudVital.ValueChanged
        Skill(GameState.EditorIndex).Vital = nudVital.Value
    End Sub

    Private Sub NudDuration_Scroll(sender As Object, e As EventArgs) Handles nudDuration.ValueChanged
        Skill(GameState.EditorIndex).Duration = nudDuration.Value
    End Sub

    Private Sub NudInterval_Scroll(sender As Object, e As EventArgs) Handles nudInterval.ValueChanged
        Skill(GameState.EditorIndex).Interval = nudInterval.Value
    End Sub

    Private Sub NudRange_Scroll(sender As Object, e As EventArgs) Handles nudRange.ValueChanged
        Skill(GameState.EditorIndex).Range = nudRange.Value
    End Sub

    Private Sub ChkAOE_CheckedChanged(sender As Object, e As EventArgs) Handles chkAoE.CheckedChanged
        If chkAoE.Checked = False Then
            Skill(GameState.EditorIndex).IsAoE = False
        Else
            Skill(GameState.EditorIndex).IsAoE = True
        End If
    End Sub

    Private Sub NudAoE_Scroll(sender As Object, e As EventArgs) Handles nudAoE.ValueChanged
        Skill(GameState.EditorIndex).AoE = nudAoE.Value
    End Sub

    Private Sub CmbAnimCast_Scroll(sender As Object, e As EventArgs) Handles cmbAnimCast.SelectedIndexChanged
        Skill(GameState.EditorIndex).CastAnim = cmbAnimCast.SelectedIndex
    End Sub

    Private Sub CmbAnim_Scroll(sender As Object, e As EventArgs) Handles cmbAnim.SelectedIndexChanged
        Skill(GameState.EditorIndex).SkillAnim = cmbAnim.SelectedIndex
    End Sub

    Private Sub NudStun_Scroll(sender As Object, e As EventArgs) Handles nudStun.ValueChanged
        Skill(GameState.EditorIndex).StunDuration = nudStun.Value
    End Sub

    Private Sub lstIndex_Click(sender As Object, e As EventArgs) Handles lstIndex.Click
        SkillEditorInit()
    End Sub

    Private Sub BtnSave_Click(sender As Object, e As EventArgs) Handles btnSave.Click
        SkillEditorOk()
        Dispose()
    End Sub

    Private Sub BtnDelete_Click(sender As Object, e As EventArgs) Handles btnDelete.Click
        Dim tmpindex As Integer

        ClearSkill(GameState.EditorIndex)

        tmpindex = lstIndex.SelectedIndex
        lstIndex.Items.RemoveAt(GameState.EditorIndex - 1)
        lstIndex.Items.Insert(GameState.EditorIndex - 1, GameState.EditorIndex & ": " & Skill(GameState.EditorIndex).Name)
        lstIndex.SelectedIndex = tmpindex

        SkillEditorInit()
    End Sub

    Private Sub BtnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        SkillEditorCancel()
        Dispose()
    End Sub

    Private Sub frmEditor_Skill_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        nudIcon.Maximum = GameState.NumSkills
        nudAoE.Maximum = Byte.MaxValue
        nudRange.Maximum = Byte.MaxValue
        nudMap.Maximum = MAX_MAPS
        lstIndex.Items.Clear()

        ' Add the names
        For i = 1 To MAX_SKILLS
            lstIndex.Items.Add(i & ": " & Trim$(Skill(i).Name))
        Next

        cmbAnimCast.Items.Clear()
        cmbAnim.Items.Clear()
        For i = 1 To MAX_ANIMATIONS
            cmbAnimCast.Items.Add(i & ": " & Type.Animation(i).Name)
            cmbAnim.Items.Add(i & ": " & Type.Animation(i).Name)
        Next

        cmbProjectile.Items.Clear()
        For i = 1 To MAX_ANIMATIONS
            cmbProjectile.Items.Add(i & ": " & Type.Projectile(i).Name)
        Next

        cmbJob.Items.Clear()
        For i = 1 To MAX_JOBS
            cmbJob.Items.Add(i & ": " & Job(i).Name.Trim)
        Next
    End Sub

    Private Sub ChkProjectile_CheckedChanged(sender As Object, e As EventArgs) Handles chkProjectile.CheckedChanged
        If chkProjectile.Checked = False Then
            Skill(GameState.EditorIndex).IsProjectile = 0
        Else
            Skill(GameState.EditorIndex).IsProjectile = 1
        End If
    End Sub

    Private Sub ScrlProjectile_Scroll(sender As System.Object, e As System.EventArgs) Handles cmbProjectile.SelectedIndexChanged
        Skill(GameState.EditorIndex).Projectile = cmbProjectile.SelectedIndex
    End Sub

    Private Sub ChkKnockBack_CheckedChanged(sender As Object, e As EventArgs) Handles chkKnockBack.CheckedChanged
        If chkKnockBack.Checked = True Then
            Skill(GameState.EditorIndex).KnockBack = 1
        Else
            Skill(GameState.EditorIndex).KnockBack = 0
        End If
    End Sub

    Private Sub CmbKnockBackTiles_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbKnockBackTiles.SelectedIndexChanged
        Skill(GameState.EditorIndex).KnockBackTiles = cmbKnockBackTiles.SelectedIndex
    End Sub

    Private Sub frmEditor_Skill_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        SkillEditorCancel()
    End Sub

    Private Sub btnLearn_Click(sender As Object, e As EventArgs) Handles btnLearn.Click
        SendLearnSkill(GameState.EditorIndex)
    End Sub

    Private Sub nudIcon_Click(sender As Object, e As EventArgs) Handles nudIcon.Click
        Skill(GameState.EditorIndex).Icon = nudIcon.Value
        DrawIcon()
    End Sub

    Private Sub DrawIcon()
        Dim skillNum As Integer
        skillNum = nudIcon.Value

        If skillNum < 1 Or skillNum > GameState.NumItems Then
            picSprite.BackgroundImage = Nothing
            Exit Sub
        End If

        If File.Exists(IO.Path.Combine(Core.Path.Skills, skillNum & GameState.GfxExt)) Then
            picSprite.BackgroundImage = Drawing.Image.FromFile(IO.Path.Combine(Core.Path.Skills, skillNum & GameState.GfxExt))
        Else
            picSprite.BackgroundImage = Nothing
        End If
    End Sub
End Class

File: Source/Client/Forms/FrmEditor_Skill.cs
﻿using System;
using System.IO;
using System.Windows.Forms;
using Core;
using Microsoft.VisualBasic;

namespace Client
{

    internal partial class frmEditor_Skill
    {
        public frmEditor_Skill()
        {
            InitializeComponent();
        }

        protected override void WndProc(ref Message m)
        {
            const int WM_MOUSEACTIVATE = 0x0021;
            const int WM_NCHITTEST = 0x0084;

            if (m.Msg == WM_MOUSEACTIVATE)
            {
                // Immediately activate and process the click.
                m.Result = new IntPtr(1); // MA_ACTIVATE
                return;
            }
            else if (m.Msg == WM_NCHITTEST)
            {
                // Let the window know the mouse is in client area.
                m.Result = new IntPtr(1); // HTCLIENT
                return;
            }

            base.WndProc(ref m);
        }

        private void TxtName_TextChanged(object sender, EventArgs e)
        {
            int tmpindex;
            tmpindex = lstIndex.SelectedIndex;
            Core.Data.Skill[GameState.EditorIndex].Name = Strings.Trim(txtName.Text);
            lstIndex.Items.RemoveAt(GameState.EditorIndex);
            lstIndex.Items.Insert(GameState.EditorIndex, GameState.EditorIndex + 1 + ": " + Core.Data.Skill[GameState.EditorIndex].Name);
            lstIndex.SelectedIndex = tmpindex;
        }

        private void CmbType_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Type = (byte)cmbType.SelectedIndex;
        }

        private void NudMp_ValueChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].MpCost = (int)Math.Round(nudMp.Value);
        }

        private void NudLevel_ValueChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].LevelReq = (int)Math.Round(nudLevel.Value);
        }

        private void CmbAccessReq_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].AccessReq = cmbAccessReq.SelectedIndex;
        }

        private void CmbClass_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].JobReq = cmbJob.SelectedIndex;
        }

        private void NudCast_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].CastTime = (int)Math.Round(nudCast.Value);
        }

        private void NudCool_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].CdTime = (int)Math.Round(nudCool.Value);
        }

        private void NudMap_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Map = (int)Math.Round(nudMap.Value);
        }

        private void CmbDir_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Dir = (byte)cmbDir.SelectedIndex;
        }

        private void NudX_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].X = (int)Math.Round(nudX.Value);
        }

        private void NudY_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Y = (int)Math.Round(nudY.Value);
        }

        private void NudVital_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Vital = (int)Math.Round(nudVital.Value);
        }

        private void NudDuration_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Duration = (int)Math.Round(nudDuration.Value);
        }

        private void NudInterval_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Interval = (int)Math.Round(nudInterval.Value);
        }

        private void NudRange_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Range = (int)Math.Round(nudRange.Value);
        }

        private void ChkAOE_CheckedChanged(object sender, EventArgs e)
        {
            if (chkAoE.Checked == false)
            {
                Core.Data.Skill[GameState.EditorIndex].IsAoE = false;
            }
            else
            {
                Core.Data.Skill[GameState.EditorIndex].IsAoE = true;
            }
        }

        private void NudAoE_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].AoE = (int)Math.Round(nudAoE.Value);
        }

        private void CmbAnimCast_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].CastAnim = cmbAnimCast.SelectedIndex;
        }

        private void CmbAnim_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].SkillAnim = cmbAnim.SelectedIndex;
        }

        private void NudStun_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].StunDuration = (int)Math.Round(nudStun.Value);
        }

        private void lstIndex_Click(object sender, EventArgs e)
        {
            Editors.SkillEditorInit();
        }

        private void BtnSave_Click(object sender, EventArgs e)
        {
            Editors.SkillEditorOK();
            Dispose();
        }

        private void BtnDelete_Click(object sender, EventArgs e)
        {
            int tmpindex;

            Database.ClearSkill(GameState.EditorIndex);

            tmpindex = lstIndex.SelectedIndex;
            lstIndex.Items.RemoveAt(GameState.EditorIndex);
            lstIndex.Items.Insert(GameState.EditorIndex, GameState.EditorIndex + 1 + ": " + Core.Data.Skill[GameState.EditorIndex].Name);
            lstIndex.SelectedIndex = tmpindex;

            Editors.SkillEditorInit();
        }

        private void BtnCancel_Click(object sender, EventArgs e)
        {
            Editors.SkillEditorCancel();
            Dispose();
        }

        private void frmEditor_Skill_Load(object sender, EventArgs e)
        {
            nudIcon.Maximum = GameState.NumSkills;
            nudAoE.Maximum = byte.MaxValue;
            nudRange.Maximum = byte.MaxValue;
            nudMap.Maximum = Constant.MAX_MAPS;
            lstIndex.Items.Clear();

            // Add the names
            for (int i = 0; i < Constant.MAX_SKILLS; i++)
                lstIndex.Items.Add(i + 1 + ": " + Strings.Trim(Core.Data.Skill[i].Name));

            cmbAnimCast.Items.Clear();
            cmbAnim.Items.Clear();
            for (int i = 0; i < Constant.MAX_ANIMATIONS; i++)
            {
                cmbAnimCast.Items.Add(i + 1 + ": " + Core.Data.Animation[i].Name);
                cmbAnim.Items.Add(i + 1 + ": " + Core.Data.Animation[i].Name);
            }

            cmbProjectile.Items.Clear();
            for (int i = 0; i < Constant.MAX_ANIMATIONS; i++)
                cmbProjectile.Items.Add(i + 1 + ": " + Core.Data.Projectile[i].Name);

            cmbJob.Items.Clear();
            for (int i = 0; i < Constant.MAX_JOBS; i++)
                cmbJob.Items.Add(i + 1 + ": " + Core.Data.Job[i].Name.Trim());
        }

        private void ChkProjectile_CheckedChanged(object sender, EventArgs e)
        {
            if (chkProjectile.Checked == false)
            {
                Core.Data.Skill[GameState.EditorIndex].IsProjectile = 0;
            }
            else
            {
                Core.Data.Skill[GameState.EditorIndex].IsProjectile = 1;
            }
        }

        private void ScrlProjectile_Scroll(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Projectile = cmbProjectile.SelectedIndex;
        }

        private void ChkKnockBack_CheckedChanged(object sender, EventArgs e)
        {
            if (chkKnockBack.Checked == true)
            {
                Core.Data.Skill[GameState.EditorIndex].KnockBack = 1;
            }
            else
            {
                Core.Data.Skill[GameState.EditorIndex].KnockBack = 0;
            }
        }

        private void CmbKnockBackTiles_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].KnockBackTiles = (byte)cmbKnockBackTiles.SelectedIndex;
        }

        private void frmEditor_Skill_FormClosing(object sender, FormClosingEventArgs e)
        {
            Editors.SkillEditorCancel();
        }

        private void btnLearn_Click(object sender, EventArgs e)
        {
            NetworkSend.SendLearnSkill(GameState.EditorIndex);
        }

        private void nudIcon_Click(object sender, EventArgs e)
        {
            Core.Data.Skill[GameState.EditorIndex].Icon = (int)Math.Round(nudIcon.Value);
            DrawIcon();
        }

        public void DrawIcon()
        {
            int skillNum;
            skillNum = (int)Math.Round(nudIcon.Value);

            if (skillNum < 1 | skillNum > GameState.NumItems)
            {
                picSprite.BackgroundImage = null;
                return;
            }

            if (File.Exists(System.IO.Path.Combine(Core.Path.Skills, skillNum + GameState.GfxExt)))
            {
                picSprite.BackgroundImage = System.Drawing.Image.FromFile(System.IO.Path.Combine(Core.Path.Skills, skillNum + GameState.GfxExt));
            }
            else
            {
                picSprite.BackgroundImage = null;
            }
        }
    }
}